server/src/main/java/org/opensearch/index/remote/RemoteIndexPathUploader.java

```diff
@@ -0,0 +1,261 @@
+        + RemoteStoreNodeAttribute.REMOTE_STORE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEY;
+
+    private static final Logger logger = LogManager.getLogger(RemoteIndexPathUploader.class);
+
+    private final Settings settings;
+    private final boolean isRemoteDataAttributePresent;
+    private final boolean isTranslogSegmentRepoSame;
+    private final Supplier<RepositoriesService> repositoriesService;
+    private volatile TimeValue indexMetadataUploadTimeout;
+
+    private BlobStoreRepository translogRepository;
+    private BlobStoreRepository segmentRepository;
+
+    public RemoteIndexPathUploader(Settings settings, Supplier<RepositoriesService> repositoriesService, ClusterSettings clusterSettings) {
+        this.settings = Objects.requireNonNull(settings);
+        this.repositoriesService = Objects.requireNonNull(repositoriesService);
+        isRemoteDataAttributePresent = isRemoteDataAttributePresent(settings);
+        // If the remote data attributes are not present, then there is no effect of translog and segment being same or different or null.
+        isTranslogSegmentRepoSame = isTranslogSegmentRepoSame();
+        Objects.requireNonNull(clusterSettings);
+        indexMetadataUploadTimeout = clusterSettings.get(INDEX_METADATA_UPLOAD_TIMEOUT_SETTING);
+        clusterSettings.addSettingsUpdateConsumer(INDEX_METADATA_UPLOAD_TIMEOUT_SETTING, this::setIndexMetadataUploadTimeout);
+    }
+
+    @Override
+    public String getThreadpoolName() {
+        return ThreadPool.Names.GENERIC;
+    }
+
+    @Override
+    public void beforeNewIndexUpload(List<IndexMetadata> indexMetadataList, ActionListener<Void> actionListener) throws IOException {
+        if (isRemoteDataAttributePresent == false) {
+            logger.trace("Skipping beforeNewIndexUpload as there are no remote indexes");
+            actionListener.onResponse(null);
+            return;
+        }
+
+        long startTime = System.nanoTime();
+        boolean success = false;
+        try {
+            List<IndexMetadata> eligibleList = indexMetadataList.stream().filter(this::requiresPathUpload).collect(Collectors.toList());
+            int latchCount = eligibleList.size() * (isTranslogSegmentRepoSame ? 1 : 2);
+            CountDownLatch latch = new CountDownLatch(latchCount);
+            List<Exception> exceptionList = Collections.synchronizedList(new ArrayList<>(latchCount));
+            for (IndexMetadata indexMetadata : eligibleList) {
+                writeIndexPathAsync(indexMetadata, latch, exceptionList);
+            }
+            String indexNames = eligibleList.stream().map(IndexMetadata::getIndex).map(Index::toString).collect(Collectors.joining(","));
+            logger.trace(new ParameterizedMessage("Remote index path upload started for {}", indexNames));
+
+            try {
+                if (latch.await(indexMetadataUploadTimeout.millis(), TimeUnit.MILLISECONDS) == false) {
```

Comment:
What happens if latch times out while upload is still in progress? Is the upload retried?

Reply:
No, there are no external retries.

Reply:
today if it times out it will fail the cluster state upload, so overall cluster state update will be retried. @ashking94 there would be internal retries for s3 SDK right?

Reply:
there can be internal retries from the sdk depending on the remote store being used.